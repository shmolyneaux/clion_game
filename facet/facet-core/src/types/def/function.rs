use crate::Shape;

/// Common fields for function pointer types
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
#[repr(C)]
#[non_exhaustive]
pub struct FunctionPointerDef {
    /// The calling abi of the function pointer
    pub abi: FunctionAbi,

    /// All parameter types, in declaration order
    pub parameters: &'static [fn() -> &'static Shape<'static>],

    /// The return type
    pub return_type: fn() -> &'static Shape<'static>,
}

/// The calling ABI of a function pointer
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, Default)]
#[repr(C)]
#[non_exhaustive]
pub enum FunctionAbi {
    /// C ABI
    C,

    /// Rust ABI
    #[default]
    Rust,

    /// An unknown ABI
    Unknown,
}
impl FunctionAbi {
    /// Returns the string in `extern "abi-string"` if not [`FunctionAbi::Unknown`].
    pub fn as_abi_str(&self) -> Option<&str> {
        match self {
            FunctionAbi::C => Some("C"),
            FunctionAbi::Rust => Some("Rust"),
            FunctionAbi::Unknown => None,
        }
    }
}

impl FunctionPointerDef {
    /// Returns a builder for FunctionPointerDef
    pub const fn builder() -> FunctionPointerDefBuilder {
        FunctionPointerDefBuilder::new()
    }
}

/// Builder for FunctionPointerDef
pub struct FunctionPointerDefBuilder {
    abi: Option<FunctionAbi>,
    parameters: &'static [fn() -> &'static Shape<'static>],
    return_type: Option<fn() -> &'static Shape<'static>>,
}

impl FunctionPointerDefBuilder {
    /// Creates a new FunctionPointerDefBuilder
    #[allow(clippy::new_without_default)]
    pub const fn new() -> Self {
        Self {
            parameters: &[],
            abi: None,
            return_type: None,
        }
    }

    /// Sets the abi for the FunctionPointerDef
    pub const fn abi(mut self, abi: FunctionAbi) -> Self {
        self.abi = Some(abi);
        self
    }

    /// Sets the parameters for the FunctionPointerDef
    pub const fn parameter_types(
        mut self,
        parameters: &'static [fn() -> &'static Shape<'static>],
    ) -> Self {
        self.parameters = parameters;
        self
    }

    /// Sets the return type for the FunctionPointerDef
    pub const fn return_type(mut self, ty: fn() -> &'static Shape<'static>) -> Self {
        self.return_type = Some(ty);
        self
    }

    /// Builds the FunctionPointerDef
    pub const fn build(self) -> FunctionPointerDef {
        FunctionPointerDef {
            parameters: self.parameters,
            return_type: self.return_type.unwrap(),
            abi: self.abi.unwrap(),
        }
    }
}
