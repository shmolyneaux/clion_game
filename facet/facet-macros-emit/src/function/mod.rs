mod emit;
pub use emit::*;
