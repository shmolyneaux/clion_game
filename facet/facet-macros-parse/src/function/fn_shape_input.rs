use proc_macro2::TokenStream;
use unsynn::*;

// Re-use the generics parser from our other module
use crate::generics::GenericParams;

unsynn! {
    /// Input to fn_shape! macro: function_name or function_name<generics>
    pub struct FnShapeInput {
        /// Function name
        pub name: Ident,
        /// Optional generic parameters
        pub generics: Option<GenericParams>,
    }
}

/// Parsed fn_shape input with extracted components
pub struct ParsedFnShapeInput {
    /// Function name
    pub name: Ident,
    /// Optional generic type parameters
    pub generics: Option<TokenStream>,
}

/// Parse fn_shape! macro input from TokenStream
pub fn parse_fn_shape_input(input: TokenStream) -> ParsedFnShapeInput {
    let mut it = input.to_token_iter();

    match it.parse::<FnShapeInput>() {
        Ok(shape_input) => {
            let name = shape_input.name;
            let generics = shape_input.generics.map(|g| g.to_token_stream());

            ParsedFnShapeInput { name, generics }
        }
        Err(err) => {
            panic!("Failed to parse fn_shape input: {}", err);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn test_simple_function_name() {
        let input = quote! { add };
        let parsed = parse_fn_shape_input(input);
        assert_eq!(parsed.name.to_string(), "add");
        assert!(parsed.generics.is_none());
    }

    #[test]
    fn test_function_with_single_generic() {
        let input = quote! { generic_add<T> };
        let parsed = parse_fn_shape_input(input);
        assert_eq!(parsed.name.to_string(), "generic_add");
        assert!(parsed.generics.is_some());
        assert_eq!(parsed.generics.unwrap().to_string().trim(), "< T >");
    }

    #[test]
    fn test_function_with_multiple_generics() {
        let input = quote! { multi_ty_param_fn<T, U> };
        let parsed = parse_fn_shape_input(input);
        assert_eq!(parsed.name.to_string(), "multi_ty_param_fn");
        assert!(parsed.generics.is_some());
        assert_eq!(parsed.generics.unwrap().to_string().trim(), "< T , U >");
    }

    #[test]
    fn test_function_with_bounded_generics() {
        let input = quote! { bounded_fn<U: Clone> };
        let parsed = parse_fn_shape_input(input);
        assert_eq!(parsed.name.to_string(), "bounded_fn");
        assert!(parsed.generics.is_some());
        assert_eq!(parsed.generics.unwrap().to_string().trim(), "< U : Clone >");
    }

    #[test]
    fn test_function_with_complex_generics() {
        let input = quote! { nested_fn<T: Add<Output = T>> };
        let parsed = parse_fn_shape_input(input);
        assert_eq!(parsed.name.to_string(), "nested_fn");
        assert!(parsed.generics.is_some());
        assert_eq!(
            parsed.generics.unwrap().to_string().trim(),
            "< T : Add < Output = T > >"
        );
    }
}
