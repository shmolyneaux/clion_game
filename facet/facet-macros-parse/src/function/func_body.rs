use crate::VerbatimUntil;
#[cfg(test)]
use proc_macro2::TokenStream;
use unsynn::*;

unsynn! {
    /// A function signature followed by its body
    pub struct FunctionWithBody {
        /// Everything before the function body (return type, etc.) - optional
        pub prefix: Option<VerbatimUntil<BraceGroup>>,
        /// The function body in braces
        pub body: BraceGroup,
    }
}

/// Parse function body from tokens using declarative parsing
/// Returns the function body as TokenStream
#[cfg(test)]
pub fn parse_function_body(tokens: &[TokenTree]) -> TokenStream {
    // Convert tokens to TokenStream for parsing
    let mut token_stream = TokenStream::new();
    for token in tokens {
        token_stream.extend(core::iter::once(token.clone()));
    }

    let mut it = token_stream.to_token_iter();

    match it.parse::<FunctionWithBody>() {
        Ok(func_with_body) => func_with_body.body.to_token_stream(),
        Err(_) => {
            // No function body found, return empty braces
            quote::quote! { {} }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn test_simple_function_body() {
        let input: Vec<TokenTree> = quote! { { x + y } }.into_iter().collect();
        let body = parse_function_body(&input);
        assert_eq!(body.to_string().trim(), "{ x + y }");
    }

    #[test]
    fn test_complex_function_body() {
        let input: Vec<TokenTree> = quote! { {
            let result = x + y;
            println!("Result: {}", result);
            result
        } }
        .into_iter()
        .collect();
        let body = parse_function_body(&input);
        let body_str = body.to_string();
        assert!(body_str.contains("let result"));
        assert!(body_str.contains("println !"));
        assert!(body_str.starts_with('{'));
        assert!(body_str.ends_with('}'));
    }

    #[test]
    fn test_function_with_trailing_tokens() {
        // Simulate tokens like: -> i32 { x + y }
        let input: Vec<TokenTree> = quote! { -> i32 { x + y } }.into_iter().collect();
        let body = parse_function_body(&input);
        assert_eq!(body.to_string().trim(), "{ x + y }");
    }

    #[test]
    fn test_no_function_body() {
        let input: Vec<TokenTree> = quote! { -> i32 }.into_iter().collect();
        let body = parse_function_body(&input);
        assert_eq!(body.to_string().trim(), "{ }");
    }

    #[test]
    fn test_nested_braces() {
        let input: Vec<TokenTree> = quote! { {
            if condition {
                inner_block();
            }
        } }
        .into_iter()
        .collect();
        let body = parse_function_body(&input);
        let body_str = body.to_string();
        assert!(body_str.contains("if condition"));
        assert!(body_str.contains("inner_block"));
    }
}
