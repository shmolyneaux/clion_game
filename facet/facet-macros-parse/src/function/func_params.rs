use crate::VerbatimUntil;
use proc_macro2::TokenStream;
use unsynn::*;

unsynn! {
    /// A function parameter with name and type
    pub struct Parameter {
        /// Parameter name
        pub name: Ident,
        /// Colon separator
        pub _colon: Colon,
        /// Parameter type (everything until comma or end)
        pub param_type: VerbatimUntil<Comma>,
    }
}

impl Parameter {
    /// Convert the parameter type to TokenStream for use with quote!
    pub fn param_type_tokens(&self) -> TokenStream {
        self.param_type.to_token_stream()
    }
}

/// Parse function parameters from a TokenStream (content of parentheses)
/// Returns a Vec of Parameter structs
pub fn parse_fn_parameters(params_ts: TokenStream) -> Vec<Parameter> {
    let mut it = params_ts.to_token_iter();

    // Parse as comma-delimited list of parameters
    match it.parse::<CommaDelimitedVec<Parameter>>() {
        Ok(params) => params.0.into_iter().map(|delim| delim.value).collect(),
        Err(_) => Vec::new(), // Empty parameter list
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn test_empty_parameters() {
        let input = quote! {};
        let params = parse_fn_parameters(input);
        assert_eq!(params.len(), 0);
    }

    #[test]
    fn test_single_parameter() {
        let input = quote! { x: i32 };
        let params = parse_fn_parameters(input);
        assert_eq!(params.len(), 1);
        assert_eq!(params[0].name.to_string(), "x");
        assert_eq!(params[0].param_type.tokens_to_string().trim(), "i32");
    }

    #[test]
    fn test_multiple_parameters() {
        let input = quote! { x: i32, y: String };
        let params = parse_fn_parameters(input);
        assert_eq!(params.len(), 2);
        assert_eq!(params[0].name.to_string(), "x");
        assert_eq!(params[0].param_type.tokens_to_string().trim(), "i32");
        assert_eq!(params[1].name.to_string(), "y");
        assert_eq!(params[1].param_type.tokens_to_string().trim(), "String");
    }

    #[test]
    fn test_complex_types() {
        let input = quote! { callback: fn(i32) -> String, data: Vec<Option<u64>> };
        let params = parse_fn_parameters(input);
        assert_eq!(params.len(), 2);
        assert_eq!(params[0].name.to_string(), "callback");
        assert_eq!(
            params[0].param_type.tokens_to_string().trim(),
            "fn (i32) -> String"
        );
        assert_eq!(params[1].name.to_string(), "data");
        assert_eq!(
            params[1].param_type.tokens_to_string().trim(),
            "Vec < Option < u64 > >"
        );
    }
}
