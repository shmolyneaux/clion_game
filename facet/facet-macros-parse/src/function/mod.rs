/// Parsing the `fn_shape!`` macro input
pub mod fn_shape_input;
/// Parsing the function body
pub mod func_body;
/// Parsing function parameters
pub mod func_params;
/// Parsing the function signature
pub mod func_sig;
/// Parsing generic functions
pub mod generics;
/// Parsing the return type
pub mod ret_type;
/// Parsing type parameters
pub mod type_params;

pub use fn_shape_input::*;
pub use func_body::*;
pub use func_params::*;
pub use func_sig::*;
pub use generics::*;
pub use ret_type::*;
pub use type_params::*;
