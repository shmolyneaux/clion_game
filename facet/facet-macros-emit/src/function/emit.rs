use facet_macros_parse::function::*;
use facet_macros_parse::{FunctionSignature, Ident, Span, TokenStream, extract_type_params};
use quote::quote;

/// Entry point for the facet_fn attribute macro
pub fn facet_fn(_attr: TokenStream, item: TokenStream) -> TokenStream {
    // Convert to proc_macro2 for parsing (quote requires it)
    let parsed = parse_function_signature(item);
    let result = generate_function_shape(parsed);

    // Convert back
    result.to_string().parse().unwrap()
}

/// Entry point for the fn_shape procedural macro  
pub fn fn_shape(input: TokenStream) -> TokenStream {
    let parsed = parse_fn_shape_input(input);
    let result = generate_fn_shape_call(parsed);

    result.to_string().parse().unwrap()
}

pub fn generate_function_shape(parsed: FunctionSignature) -> TokenStream {
    let fn_name = parsed.name;
    let generics = parsed.generics;
    let params = parsed.parameters;
    let return_type = parsed.return_type;
    let body = parsed.body;

    let hidden_mod = Ident::new(&format!("__fn_shape_{}", fn_name), Span::call_site());
    let shape_name = Ident::new(
        &format!("{}_SHAPE", fn_name.to_string().to_uppercase()),
        Span::call_site(),
    );
    let defs: Vec<_> = params
        .iter()
        .map(|p| {
            let name = &p.name;
            let ty = &p.param_type_tokens();
            quote! { #name: #ty }
        })
        .collect();
    let idents: Vec<_> = params
        .iter()
        .map(|p| {
            let name = &p.name;
            quote! { #name }
        })
        .collect();
    let types: Vec<_> = params
        .iter()
        .map(|p| {
            let ty = &p.param_type_tokens();
            quote! { #ty }
        })
        .collect();
    let names: Vec<_> = params
        .iter()
        .map(|p| p.name.to_string())
        .collect::<Vec<_>>();
    let arity = params.len();
    let fn_name_str = fn_name.to_string();

    // Extract  type parameters for PhantomData using unsynn parsing
    let generics_type = if let Some(ref generics_ts) = generics {
        extract_type_params(generics_ts.clone())
    } else {
        quote! { () }
    };

    let documentation_lines: Vec<_> = parsed
        .documentation
        .iter()
        .map(|doc| quote! { #doc })
        .collect();

    let shape_definition = quote! {
        pub fn shape #generics () -> FunctionShape<( #( #types ),* ), #return_type, #generics_type> {
            FunctionShape::new(
                #fn_name_str,
                #arity,
                &[ #( #names ),* ],
                &[ #( #documentation_lines ),* ]
            )
        }
    };

    let out = quote! {
        // 1) Move the real implementation into a private module
        #[allow(non_snake_case)]
        mod #hidden_mod {
            use super::*;
            pub(super) fn inner #generics ( #( #defs ),* ) -> #return_type #body

            #[derive(Debug, Clone)]
            pub struct FunctionShape<Args, Ret, Generics = ()> {
                pub name: &'static str,
                pub param_count: usize,
                pub param_names: &'static [&'static str],
                pub documentation: &'static [&'static str],
                _args: core::marker::PhantomData<Args>,
                _ret: core::marker::PhantomData<Ret>,
                _generics: core::marker::PhantomData<Generics>,
            }

            impl<Args, Ret, Generics> FunctionShape<Args, Ret, Generics> {
                pub const fn new(
                    name: &'static str,
                    param_count: usize,
                    param_names: &'static [&'static str],
                    documentation: &'static [&'static str],
                ) -> Self {
                    Self {
                        name,
                        param_count,
                        param_names,
                        documentation,
                        _args: core::marker::PhantomData,
                        _ret: core::marker::PhantomData,
                        _generics: core::marker::PhantomData,
                    }
                }
            }

            #shape_definition
        }

        // 2) Public wrapper retains the exact original signature
        pub fn #fn_name #generics ( #( #defs ),* ) -> #return_type {
            #hidden_mod::inner( #( #idents ),* )
        }

        // 3) Re-export the shape function with function name
        pub use #hidden_mod::shape as #shape_name;
    };

    out
}

fn generate_fn_shape_call(parsed: ParsedFnShapeInput) -> TokenStream {
    let fn_name = parsed.name;
    let generic_args = parsed.generics;

    // Generate the shape function name
    let shape_name = Ident::new(
        &format!("{}_SHAPE", fn_name.to_string().to_uppercase()),
        Span::call_site(),
    );

    if let Some(generics) = generic_args {
        quote! { #shape_name::#generics() }
    } else {
        quote! { #shape_name() }
    }
}
