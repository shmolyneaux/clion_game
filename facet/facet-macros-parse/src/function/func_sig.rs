use proc_macro2::TokenStream;
use unsynn::*;

// Re-use the types from our other modules
use crate::func_params::Parameter;
use crate::generics::GenericParams;
use crate::ret_type::ReturnType;
use crate::{Attribute, AttributeInner};

keyword! {
    /// The "fn" keyword.
    pub KFn = "fn";
}

// We need to define how to parse different types of attributes
unsynn! {
    /// A complete function signature with optional attributes
    pub struct FnSignature {
        /// Zero or more attributes (including doc comments)
        pub attributes: Option<Many<Attribute>>,
        /// The "fn" keyword
        pub _fn_keyword: KFn,
        /// Function name
        pub name: Ident,
        /// Optional generic parameters
        pub generics: Option<GenericParams>,
        /// Parameter list in parentheses
        pub params: ParenthesisGroup,
        /// Optional return type
        pub return_type: Option<ReturnType>,
        /// Function body
        pub body: BraceGroup,
    }
}

/// Parsed function signature with extracted components including documentation
pub struct FunctionSignature {
    /// Function name
    pub name: Ident,
    /// Optional generic parameters
    pub generics: Option<TokenStream>,
    /// Function parameters
    pub parameters: Vec<Parameter>,
    /// Optional return type
    pub return_type: TokenStream,
    /// Function body
    pub body: TokenStream,
    /// Extracted documentation from doc comments as separate lines
    pub documentation: Vec<String>,
}

/// Extract documentation from attributes
pub fn extract_documentation(attributes: &Option<Many<Attribute>>) -> Vec<String> {
    let attrs = match attributes {
        Some(many_attrs) => &many_attrs.0,
        None => return Vec::new(),
    };

    let mut doc_lines = Vec::new();

    for attr in attrs.iter() {
        // Pattern match on the AttributeInner enum
        match &attr.value.body.content {
            AttributeInner::Doc(doc_attr) => {
                // LiteralString lets you access the unquoted value, but must be unescaped
                let doc_str = doc_attr.value.as_str().replace("\\\"", "\"");
                doc_lines.push(doc_str);
            }
            AttributeInner::Facet(_) | AttributeInner::Repr(_) | AttributeInner::Any(_) => {
                // Not a doc attribute, skipping
            }
        }
    }

    doc_lines
}

/// Parse a complete function signature from TokenStream
pub fn parse_function_signature(input: TokenStream) -> FunctionSignature {
    let mut it = input.to_token_iter();

    match it.parse::<FnSignature>() {
        Ok(sig) => {
            // Extract parameters from the parenthesis group
            let params_content = {
                let params_tokens = sig.params.to_token_stream();
                let mut it = params_tokens.to_token_iter();
                if let Ok(TokenTree::Group(group)) = it.parse::<TokenTree>() {
                    group.stream()
                } else {
                    TokenStream::new()
                }
            };
            let parameters = crate::func_params::parse_fn_parameters(params_content);

            // Extract generics if present
            let generics = sig.generics.map(|g| g.to_token_stream());

            // Extract return type if present
            let return_type = sig
                .return_type
                .map(|rt| rt.return_type.to_token_stream())
                .unwrap_or_else(|| quote::quote! { () });

            // Extract body
            let body = sig.body.to_token_stream();

            // Extract documentation from attributes
            let documentation = extract_documentation(&sig.attributes);

            FunctionSignature {
                name: sig.name,
                generics,
                parameters,
                return_type,
                body,
                documentation,
            }
        }
        Err(err) => {
            panic!("Failed to parse function signature: {}", err);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn test_simple_function() {
        let input = quote! {
            fn add(x: i32, y: i32) -> i32 {
                x + y
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "add");
        assert!(parsed.generics.is_none());
        assert_eq!(parsed.parameters.len(), 2);
        assert_eq!(parsed.parameters[0].name.to_string(), "x");
        assert_eq!(parsed.parameters[1].name.to_string(), "y");
        assert_eq!(parsed.return_type.to_string().trim(), "i32");
    }

    #[test]
    fn test_generic_function() {
        let input = quote! {
            fn generic_add<T>(x: T, y: T) -> T {
                x + y
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "generic_add");
        assert!(parsed.generics.is_some());
        assert_eq!(parsed.parameters.len(), 2);
        assert_eq!(parsed.return_type.to_string().trim(), "T");
    }

    #[test]
    fn test_no_params_function() {
        let input = quote! {
            fn no_params() -> &'static str {
                "hello"
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "no_params");
        assert_eq!(parsed.parameters.len(), 0);
        assert_eq!(parsed.return_type.to_string().trim(), "& 'static str");
    }

    #[test]
    fn test_no_return_type() {
        let input = quote! {
            fn no_return(x: i32) {
                println!("{}", x);
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "no_return");
        assert_eq!(parsed.parameters.len(), 1);
        assert_eq!(parsed.return_type.to_string().trim(), "()");
    }

    #[test]
    fn test_function_with_doc_comments() {
        let input = quote! {
            #[doc = " This is a test function"]
            #[doc = " that does addition of two numbers"]
            fn add(x: i32, y: i32) -> i32 {
                x + y
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "add");
        assert!(parsed.generics.is_none());
        assert_eq!(parsed.parameters.len(), 2);
        assert_eq!(parsed.parameters[0].name.to_string(), "x");
        assert_eq!(parsed.parameters[1].name.to_string(), "y");
        assert_eq!(parsed.return_type.to_string().trim(), "i32");

        // Check documentation
        assert!(!parsed.documentation.is_empty());
        assert_eq!(parsed.documentation.len(), 2); // Two doc lines
        assert_eq!(parsed.documentation[0], " This is a test function");
        assert_eq!(
            parsed.documentation[1],
            " that does addition of two numbers"
        );
    }

    #[test]
    fn test_function_with_single_doc_comment() {
        let input = quote! {
            #[doc = " Single line documentation"]
            fn greet(name: String) -> String {
                format!("Hello, {}!", name)
            }
        };

        let parsed = parse_function_signature(input);
        assert!(!parsed.documentation.is_empty());
        assert_eq!(parsed.documentation.len(), 1);
        assert_eq!(parsed.documentation[0], " Single line documentation");
    }

    #[test]
    fn test_function_without_doc_comments() {
        let input = quote! {
            fn no_docs(x: i32) -> i32 {
                x * 2
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "no_docs");
        assert!(parsed.documentation.is_empty());
    }

    #[test]
    fn test_function_with_mixed_attributes() {
        let input = quote! {
            #[doc = " Documentation comment"]
            #[derive(Debug)]
            #[doc = " More documentation"]
            fn mixed_attrs() {
                println!("test");
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "mixed_attrs");
        assert!(!parsed.documentation.is_empty());
        assert_eq!(parsed.documentation.len(), 2); // Two doc lines

        assert_eq!(parsed.documentation[0], " Documentation comment");
        assert_eq!(parsed.documentation[1], " More documentation");
    }

    #[test]
    fn test_generic_function_with_docs() {
        // Imitate `///d` with `#[doc = "d"]` because `quote!` gives `r"d"` not `"d"`
        // which `unsynn::LiteralString` will not match on when parsing.
        let input = quote! {
            #[doc = " Generic function that adds two values"]
            fn generic_add<T: Add<Output = T>>(x: T, y: T) -> T {
                x + y
            }
        };

        let parsed = parse_function_signature(input);
        assert_eq!(parsed.name.to_string(), "generic_add");
        assert!(parsed.generics.is_some());
        assert!(!parsed.documentation.is_empty());
        assert_eq!(parsed.documentation.len(), 1);

        // Check the content
        assert_eq!(
            parsed.documentation[0],
            " Generic function that adds two values"
        );
    }
}
