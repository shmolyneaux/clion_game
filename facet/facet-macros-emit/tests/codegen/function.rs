use facet_macros_emit::function::generate_function_shape;
use facet_macros_parse::TokenStream;
use facet_macros_parse::function::*;
use rust_format::{Formatter, RustFmt};

/// Helper function to parse and generate function shape code, then format it
fn expand_function(input: &str) -> String {
    let trimmed_input = input.trim();
    let input_ts: TokenStream = trimmed_input.parse().expect("Failed to parse input");

    let parsed = parse_function_signature(input_ts);
    let generated = generate_function_shape(parsed);

    RustFmt::default()
        .format_tokens(generated)
        .unwrap_or_else(|e| panic!("Format error: {e}"))
}

/// Helper function to parse function signature and verify its components
fn parse_and_verify_signature(input: &str) -> FunctionSignature {
    let trimmed_input = input.trim();
    let input_ts: TokenStream = trimmed_input.parse().expect("Failed to parse input");
    parse_function_signature(input_ts)
}

#[test]
fn test_simple_function_parsing() {
    let input = r#"
        fn add(x: i32, y: i32) -> i32 {
            x + y
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "add");
    assert!(parsed.generics.is_none());
    assert_eq!(parsed.parameters.len(), 2);
    assert_eq!(parsed.parameters[0].name.to_string(), "x");
    assert_eq!(parsed.parameters[1].name.to_string(), "y");
    assert_eq!(parsed.return_type.to_string().trim(), "i32");
}

#[test]
#[ignore]
fn test_no_params_function_parsing() {
    let input = r#"
        fn no_params() -> &'static str {
            "No parameters here!"
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "no_params");
    assert!(parsed.generics.is_none());
    assert_eq!(parsed.parameters.len(), 0);
    assert_eq!(parsed.return_type.to_string().trim(), "& 'static str");
}

#[test]
fn test_generic_function_parsing() {
    let input = r#"
        fn generic_add<T: Add<Output = T>>(x: T, y: T) -> T {
            x + y
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "generic_add");
    assert!(parsed.generics.is_some());
    assert_eq!(parsed.parameters.len(), 2);
    assert_eq!(parsed.parameters[0].name.to_string(), "x");
    assert_eq!(parsed.parameters[1].name.to_string(), "y");
    assert_eq!(parsed.return_type.to_string().trim(), "T");
}

#[test]
fn test_string_function_parsing() {
    let input = r#"
        fn greet(name: String) -> String {
            format!("Hello, {}!", name)
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "greet");
    assert!(parsed.generics.is_none());
    assert_eq!(parsed.parameters.len(), 1);
    assert_eq!(parsed.parameters[0].name.to_string(), "name");
    assert_eq!(parsed.return_type.to_string().trim(), "String");
}

#[test]
fn test_complex_generic_function_parsing() {
    let input = r#"
        fn complex_fn<T, U>(x: Vec<T>, y: Option<U>) -> Result<T, U> {
            todo!()
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "complex_fn");
    assert!(parsed.generics.is_some());
    assert_eq!(parsed.parameters.len(), 2);
    assert_eq!(parsed.parameters[0].name.to_string(), "x");
    assert_eq!(parsed.parameters[1].name.to_string(), "y");
    assert_eq!(parsed.return_type.to_string().trim(), "Result < T , U >");
}

#[test]
fn test_no_return_type_function_parsing() {
    let input = r#"
        fn no_return(x: i32) {
            println!("{}", x);
        }
    "#;

    let parsed = parse_and_verify_signature(input);
    assert_eq!(parsed.name.to_string(), "no_return");
    assert!(parsed.generics.is_none());
    assert_eq!(parsed.parameters.len(), 1);
    assert_eq!(parsed.parameters[0].name.to_string(), "x");
    assert_eq!(parsed.return_type.to_string().trim(), "()");
}

#[test]
fn simple_function() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn add(x: i32, y: i32) -> i32 {
            x + y
        }
        "#
    ));
}

#[test]
fn function_with_string_parameter() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn greet(name: String) -> String {
            format!("Hello, {}!", name)
        }
        "#
    ));
}

#[test]
fn function_with_no_parameters() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn no_params() -> &'static str {
            "No parameters here!"
        }
        "#
    ));
}

#[test]
fn function_with_no_return_type() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn no_return(x: i32) {
            println!("{}", x);
        }
        "#
    ));
}

#[test]
fn generic_function_simple() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn generic_add<T: Add<Output = T>>(x: T, y: T) -> T {
            x + y
        }
        "#
    ));
}

#[test]
fn generic_function_multiple_params() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn complex_fn<T, U>(x: Vec<T>, y: Option<U>) -> Result<T, U> {
            todo!()
        }
        "#
    ));
}

#[test]
fn generic_function_with_complex_bounds() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn bounded_fn<T: Clone + Send, U: Iterator<Item = T>>(data: U) -> Vec<T> {
            data.collect()
        }
        "#
    ));
}

#[test]
#[ignore]
fn function_with_complex_types() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn complex_types(
            callback: fn(i32) -> String, 
            data: Vec<Option<u64>>,
            result: Result<String, Box<dyn std::error::Error>>
        ) -> HashMap<String, Vec<i32>> {
            todo!()
        }
        "#
    ));
}

#[test]
#[ignore]
fn function_with_references_and_lifetimes() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn with_lifetimes<'a>(s: &'a str, data: &'a [u8]) -> &'a str {
            s
        }
        "#
    ));
}

#[test]
fn function_with_mutable_references() {
    insta::assert_snapshot!(expand_function(
        r#"
        fn with_mut_refs(x: &mut i32, y: &mut Vec<String>) -> usize {
            *x += 1;
            y.len()
        }
        "#
    ));
}

#[test]
fn function_with_single_doc_comment() {
    insta::assert_snapshot!(expand_function(
        r#"
        /// Single line documentation
        fn greet(name: String) -> String {
            format!("Hello, {}!", name)
        }
        "#
    ));
}

#[test]
fn function_with_multiple_doc_comments() {
    insta::assert_snapshot!(expand_function(
        r#"
        /// This is a test function
        /// that does addition of two numbers
        fn add(x: i32, y: i32) -> i32 {
            x + y
        }
        "#
    ));
}

#[test]
fn function_with_doc_comments_and_quotes() {
    insta::assert_snapshot!(expand_function(
        r#"
        /// Hello "world", if that is your real name
        fn greet(name: String) -> String {
            format!("Hello, {}...?", name)
        }
        "#
    ));
}

#[test]
fn function_with_complex_doc_comments() {
    insta::assert_snapshot!(expand_function(
        r###"
        /// This uses r#"raw strings"# and r##"nested"## syntax
        fn complex_doc() {
            println!("test");
        }
        "###
    ));
}

#[test]
fn function_with_mixed_attributes() {
    insta::assert_snapshot!(expand_function(
        r#"
        /// Documentation comment
        #[derive(Debug)]
        /// More documentation
        fn mixed_attrs() {
            println!("test");
        }
        "#
    ));
}

#[test]
fn generic_function_with_docs() {
    insta::assert_snapshot!(expand_function(
        r#"
        /// Generic function that adds two values
        fn generic_add<T: Add<Output = T>>(x: T, y: T) -> T {
            x + y
        }
        "#
    ));
}
