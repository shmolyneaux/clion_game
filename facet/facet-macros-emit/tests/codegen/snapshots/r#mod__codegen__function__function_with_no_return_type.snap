---
source: facet-macros-emit/tests/codegen/function.rs
expression: "expand_function(r#\"\n        fn no_return(x: i32) {\n            println!(\"{}\", x);\n        }\n        \"#)"
---
#[allow(non_snake_case)]
mod __fn_shape_no_return {
    use super::*;
    pub(super) fn inner(x: i32) -> () {
        println!("{}", x);
    }
    #[derive(Debug, Clone)]
    pub struct FunctionShape<Args, Ret, Generics = ()> {
        pub name: &'static str,
        pub param_count: usize,
        pub param_names: &'static [&'static str],
        pub documentation: &'static [&'static str],
        _args: core::marker::PhantomData<Args>,
        _ret: core::marker::PhantomData<Ret>,
        _generics: core::marker::PhantomData<Generics>,
    }
    impl<Args, Ret, Generics> FunctionShape<Args, Ret, Generics> {
        pub const fn new(
            name: &'static str,
            param_count: usize,
            param_names: &'static [&'static str],
            documentation: &'static [&'static str],
        ) -> Self {
            Self {
                name,
                param_count,
                param_names,
                documentation,
                _args: core::marker::PhantomData,
                _ret: core::marker::PhantomData,
                _generics: core::marker::PhantomData,
            }
        }
    }
    pub fn shape() -> FunctionShape<(i32), (), ()> {
        FunctionShape::new("no_return", 1usize, &["x"], &[])
    }
}
pub fn no_return(x: i32) -> () {
    __fn_shape_no_return::inner(x)
}
pub use __fn_shape_no_return::shape as NO_RETURN_SHAPE;
